File: 4-Behavioral Patterns/10-Observer Pattern/Tweeter-Follower Example Without Builtin/Java/runner/ObserverPatternTest.java
package runner;

import observer.Follower;
import subject.Tweeter;

/**
 * Application that actually uses Observer Pattern.
 *
 * @author Ziang Lu
 */
public class ObserverPatternTest {

    /**
     * Main driver.
     * @param args arguments from command line
     */
    public static void main(String[] args) {
        Tweeter trump = new Tweeter("Donald Trump");
        Follower american1 = new Follower("US Citizen 1");
        Follower american2 = new Follower("US Citizen 2");
        american1.follow(trump);
        american2.follow(trump);

        System.out.println();
        trump.writeTweeter();

        /*
         * Output:
         * US Citizen 1: Starting to follow Donald Trump
         * US Citizen 2: Starting to follow Donald Trump
         *
         * Donald Trump: I wrote a new tweet: A wonderful way with my family!
         * US Citizen 2: I got an update from Donald Trump about a new tweet: A wonderful way with my family!
         * US Citizen 1: I got an update from Donald Trump about a new tweet: A wonderful way with my family!
         */
    }

}

File: 4-Behavioral Patterns/11-Observer Pattern/Tweeter-Follower Example Without Builtin/Python/observer_pattern_test.py
#!usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Application that actually uses Observer Pattern.
"""

__author__ = 'Ziang Lu'

from observer import Follower
from subject import Tweeter


def main():
    trump = Tweeter('Donald Trump')
    american1 = Follower('US Citizen 1')
    american2 = Follower('US Citizen 2')
    american1.follow(trump)
    american2.follow(trump)

    print()
    trump.write_tweet('A wonderful day with my family!')

    print()
    american1.unfollow(trump)
    american2.unfollow(trump)


if __name__ == '__main__':
    main()

# Output:
# US Citizen 1: Starting to follow Donald Trump
# US Citizen 2: Starting to follow Donald Trump
#
# Donald Trump: I wrote a new tweet: A wonderful day with my family!
# US Citizen 2: I got an update from Donald Trump about a new tweet: A wonderful day with my family!
# US Citizen 1: I got an update from Donald Trump about a new tweet: A wonderful day with my family!
#
# US Citizen 1: Stops to follow Donald Trump
# US Citizen 2: Stops to follow Donald Trump
